app/upgrades/v1-fork-btc-reorg-k/testnet/upgrade.go

```diff
@@ -0,0 +1,353 @@
+	for btcHeight := btcBlockHeightRollbackFrom; btcHeight <= btcTipHeight+higherBtcStakingPeriod; btcHeight++ {
+		vpEvents := btcStkK.GetAllPowerDistUpdateEvents(ctx, btcHeight, btcHeight)
+		for idx, evt := range vpEvents {
+			switch typedEvent := evt.Ev.(type) {
+			case *bstypes.EventPowerDistUpdate_BtcDelStateUpdate:
+				delEvt := typedEvent.BtcDelStateUpdate
+				stkTxHash := delEvt.StakingTxHash
+
+				btcDel := loadBtcDel(ctx, btcStkK, cacheBtcDelByStkTxHashHex, stkTxHash)
+
+				if largerBtcReorg.IsBtcHeightRollbacked(btcDel.StartHeight) {
+					mapStkTxHashRollback[stkTxHash] = struct{}{}
+					eventsToDelete = append(eventsToDelete, bstypes.EventIndex{
+						Idx:            uint64(idx),
+						BlockHeightBtc: btcHeight,
+					})
+					l.Debug(
+						"staking transaction was rolledbacked",
+						zap.String("stk_tx_hash_hex", stkTxHash),
+					)
+					continue
+				}
+				// if the btc staking transaction hash was not activated during the rollback
+				// verify if the unbond wasn't
+
+				if delEvt.NewState != bstypes.BTCDelegationStatus_UNBONDED {
+					// if it is not unbonding, nothing to do
+					continue
+				}
+
+				_, isUnbondingTxRolledBack := mapUnbondStkTxHashRollback[stkTxHash]
+				if isUnbondingTxRolledBack {
+					l.Debug(
+						"unbonding transaction was rolledbacked",
+						zap.String("stk_tx_hash_hex", stkTxHash),
+					)
+					eventsToDelete = append(eventsToDelete, bstypes.EventIndex{
+						Idx:            uint64(idx),
+						BlockHeightBtc: btcHeight,
+					})
+				}
+
+			default: // other events as slashed, jailed, unjail do not matter for the rollback procedure
+				continue
+			}
+		}
+	}
+
+	for _, evt := range eventsToDelete {
+		btcStkK.DeletePowerDistEvent(ctx, evt.BlockHeightBtc, evt.Idx)
+	}
+
+	return mapStkTxHashRollback
+}
+
+// HandleBtcDelegationsAndIncentive rollbacks all the BTC staking transactions that were included during the blocks
+// which were rollbacked in the BTC reorg.
+// Note: the order of the endblock which panic matters, the halt happened at the btcstaking
+// so finality and incentive had not run their end blockers in the panic block.
+func HandleBtcDelegationsAndIncentive(
+	ctx context.Context,
+	btcStkK *bskeeper.Keeper,
+	finalK *fkeeper.Keeper,
+
+	cacheBtcDelByStkTxHashHex map[string]*bstypes.BTCDelegation,
+	btcTipHeight uint32,
+	paramsByVs map[uint32]*bstypes.Params,
+	mapStkTxHashRollback, mapRollbackUnbondTxs map[string]struct{},
+) (satsToUnbondByFpBtcPk, satsToActivateByFpBtcPk map[string]uint64, err error) {
+	cacheFpByBtcPkHex := make(map[string]*bstypes.FinalityProvider, 0)
+	// collect data to update cache voting power table in finality
+	satsToUnbondByFpBtcPk, satsToActivateByFpBtcPk = make(map[string]uint64, 0), make(map[string]uint64, 0)
+
+	for stkTxHash := range mapStkTxHashRollback {
+		btcDel := loadBtcDel(ctx, btcStkK, cacheBtcDelByStkTxHashHex, stkTxHash)
+
+		if !btcDel.HasInclusionProof() {
```

Comment:
shouldn't we remove these anyways, considering these will be non-existent in BTC

Reply:
good point, maybe we should delete it entirely